src/main/java/fi/hsl/transitdata/tripupdate/processing/TripCancellationProcessor.java

```diff
@@ -47,10 +41,15 @@ public boolean validateMessage(Message msg) {
     }
 
     @Override
-    public Optional<GtfsRealtime.TripUpdate> processMessage(Message msg) {
+    public Optional<TripUpdateWithId> processMessage(Message msg) {
         try {
             InternalMessages.TripCancellation tripCancellation = InternalMessages.TripCancellation.parseFrom(msg.getData());
-            return tripUpdateProcessor.processTripCancellation(msg.getEventTime(), tripCancellation);
+            final String tripId = tripCancellation.getTripId();
+
+            Optional<GtfsRealtime.TripUpdate> maybeTripUpdate = tripUpdateProcessor.processTripCancellation(msg.getEventTime(), tripCancellation);
+            return maybeTripUpdate.flatMap(tripUpdate ->
+                    TripUpdateWithId.newInstance(tripId, tripUpdate)
+            );
```

Comment:
it flattens Optional<Optional<TripUpdate>> into Optional<TripUpdate>. And yes, tripUpdate variable cannot be null inside the lambda because then Optional<TripUpdate> would be empty and the lambda would never be called.